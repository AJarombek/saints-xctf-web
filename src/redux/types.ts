/**
 * Types for the Redux store.
 * @author Andrew Jarombek
 * @since 7/23/2020
 */

export type RootState = {
    auth: AuthState,
    logs: LogsState
}

export type LogsState = {
    isFetching: boolean;
    didInvalidate: boolean;
    lastUpdated: number;
    items: Logs;
    feeds: LogFeeds;
    newComments: NewComments;
}

export type AuthState = {
    auth: Auth;
    user: Users;
}

export interface Meta {
    isFetching?: boolean;
    lastUpdated?: number;
    didInvalidate?: boolean;
    serverError?: string;
}

export type Auth = {
    isFetching?: boolean;
    lastUpdated?: number;
    signedInUser?: string;
    status?: string;
}

export type Users = {
    [key: string]: UserMeta;
}

export interface UserMeta extends User, Meta {}

export interface User {
    activation_code?: string;
    class_year?: number;
    deleted?: string;
    description?: string;
    email?: string;
    favorite_event?: string;
    first?: string;
    last?: string;
    last_signin?: string;
    location?: string;
    member_since?: string;
    password?: string;
    salt?: string;
    subscribed?: string;
    username?: string;
    week_start?: string;
}

enum LogType { RUN = "run", BIKE = "bike", SWIM = "swim", OTHER = "other" }
enum Metric { MILES = "miles", KILOMETERS = "kilometers", METERS = "meters" }

export type Log = {
    log_id: number;
    username: string;
    first: string;
    last: string;
    name: string;
    date: string;
    type: LogType;
    distance?: number;
    metric?: Metric;
    miles?: number;
    time?: string;
    pace?: string;
    feel: number;
    location?: string;
    description?: string;
    timeCreated?: string;
    comments?: Comment[];
}

export type Logs = {
    [key: string]: Record<string, Log>;
}

export type LogFeeds = {
    [key: string]: {
        filterBy: string;
        bucket: string;
        pages: LogFeedPages
    };
}

export type LogFeedPages = {
    [key: string]: LogFeedPage
}

export type LogFeedPage = {
    isFetching: boolean;
    lastUpdated: number;
    items: Log[];
    serverError: string;
}

export type Comment = {
    comment_id: number;
    username: string;
    first: string;
    last: string;
    log_id: number;
    time: string;
    content: string;
}

export type NewComments = {
    [key: string]: {
        isFetching: boolean;
        lastUpdated: number;
        created?: boolean;
        serverError?: string;
    }
}
