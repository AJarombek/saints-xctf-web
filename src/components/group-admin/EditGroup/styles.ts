/**
 * JSS styles for the EditGroup component.
 * @author Andrew Jarombek
 * @since 1/26/2021
 */

import Mixins from '../../../styles/mixins';
import color from 'color';

export default {
  editGroup: {},
  title: {
    ...Mixins.formTitle(),
    fontSize: 20
  },
  form: {
    ...Mixins.containerBackground(),
    backgroundColor: color('#E3E3E3').alpha(0.8).string(),
    maxWidth: '900px'
  },
  inputTitle: {
    ...Mixins.inputTitle()
  },
  radioGroup: {
    display: 'flex'
  },
  radio: {
    margin: '0 20px'
  },
  textArea: {
    ...Mixins.textArea()
  },
  actions: {
    ...Mixins.formActions()
  },
  submitButton: {
    ...Mixins.submitButton()
  },
  cancelButton: {
    marginRight: 10
  },
  disabledSubmitButton: {
    ...Mixins.disabledButton(),
    backgroundColor: '#d6d6d6 !important'
  },
  buttonSpinner: {
    ...Mixins.buttonSpinner()
  }
};
