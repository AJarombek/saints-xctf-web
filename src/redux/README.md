### Overview

Configure Redux as the state management system for the application.

### Files

| Filename                 | Description                                                                |
|--------------------------|----------------------------------------------------------------------------|
| `modules`                | Redux modules containing Reducers, Actions, and Action Creators.           |
| `appState.js`            | Sample application state maintained by Redux.                              |
| `store.js`               | Configuration and creation of the Redux store.                             |
| `types.ts`               | Typescript types used throughout the application.                          |

### Resources

1) [React/Redux Example App #1](https://github.com/erikras/react-redux-universal-hot-example/tree/master/src/redux)
2) [React/Redux Example App #2](https://github.com/goopscoop/ga-react-tutorial/tree/6-reduxActionsAndReducers/src/redux)
3) [React/Redux Example App #3](https://github.com/gothinkster/react-redux-realworld-example-app/tree/master/src)
4) [Redux Ducks Spec](https://github.com/erikras/ducks-modular-redux)
5) [React Router/Redux Setup Steps](https://github.com/supasate/connected-react-router#usage)
6) [Redux Hot Reloading](https://redux.js.org/recipes/configuring-your-store/#hot-reloading)
7) [Redux Thunk](https://github.com/reduxjs/redux-thunk)
8) [Redux/React Official Real World Example](https://github.com/reduxjs/redux/tree/master/examples/real-world)
