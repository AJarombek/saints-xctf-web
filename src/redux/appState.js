/**
 * Sample Redux State for the entire application.
 * @author Andrew Jarombek
 * @since 5/3/2020
 */

const appState = {
  auth: {
    auth: {
      isFetching: false,
      lastUpdated: 1588530309,
      signedIn: true,
      status: "SUCCESS"
    },
    registration: {
      stage: 1,
      first: "Andy",
      last: "Jarombek",
      email: "andrew@jarombek.com",
      username: undefined,
      password: undefined,
      activationCode: undefined,
      teams: {
        saintsXCTFAlumni: {
          status: "accepted",
          groups: {
            alumni: {
              status: "accepted"
            }
          }
        }
      }
    },
    user: {
      andy: {
        isFetching: false,
        didInvalidate: false,
        lastUpdated: 1588530310,
        username: "andy",
        first: "Andrew",
        last: "Jarombek",
        password: "xxxxxx",
        description: "Test Description",
        memberSince: "2016-02-23 12:00:00",
        classYear: 2017,
        location: "New York, NY",
        favoriteEvent: "Cooldown Jog",
        activationCode: "ABCD1234",
        email: "andrew@jarombek.com",
        lastSignIn: "2020-05-03 12:00:00",
        subscribed: "Y",
        logs: []
      }
    },
  },
  profile: {
    tom: {
      isFetching: false,
      didInvalidate: false,
      lastUpdated: 1588530311
    },
    joe: {
      isFetching: false,
      didInvalidate: false,
      lastUpdated: 1588530312
    },
    ben: {
      isFetching: false,
      didInvalidate: false,
      lastUpdated: 1588530313
    }
  },
  group: {
    wmenstf: {
      isFetching: false,
      didInvalidate: false,
      lastUpdated: 1588530314,
      groupName: "wmenstf",
      groupTitle: "Women's Track & Field",
      description: "",
      weekStart: "monday",
      members: [],
      logs: []
    }
  },
  logs: {
    isFetching: false,
    didInvalidate: false,
    lastUpdated: 1588530315,
    items: {
      1: {
        log_id: 1,
        username: "andy",
        first: "Andrew",
        last: "Jarombek",
        name: "Reservoir Loop",
        location: "New York, NY",
        date: "2020-05-03",
        type: "run",
        distance: 6.37,
        metric: "miles",
        miles: 6.37,
        time: "00:45:37",
        pace: "00:07:10",
        feel: 6,
        description: "",
        timeCreated: "2020-05-03 12:00:00",
        comments: {
          1: {
            commentId: 1,
            username: "andy",
            first: "Andrew",
            last: "Jarombek",
            logId: 1,
            time: "2020-05-03 12:00:00",
            content: "test"
          }
        }
      }
    }
  }
};
