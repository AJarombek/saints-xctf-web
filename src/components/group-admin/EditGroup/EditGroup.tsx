/**
 * Component for editing a group as an admin.
 * @author Andrew Jarombek
 * @since 1/26/2021
 */

import React, { useEffect, useRef, useState } from 'react';
import { createUseStyles } from 'react-jss';
import styles from './styles';
import UploadGroupPicture from '../UploadGroupPicture';
import { Group, GroupMeta } from '../../../redux/types';
import AutoResizeTextArea from '../../shared/AutoResizeTextArea';
import RadioButton from '../../shared/RadioButton';
import { AJButton } from 'jarombek-react-components';
import { useDispatch } from 'react-redux';
import { putGroup } from '../../../redux/modules/groups';
import DefaultErrorPopup from '../../shared/DefaultErrorPopup';
import LoadingSpinner from '../../shared/LoadingSpinner';
import classNames from 'classnames';

interface Props {
  group: GroupMeta;
}

const useStyles = createUseStyles(styles);

const EditGroup: React.FunctionComponent<Props> = ({ group }) => {
  const classes = useStyles();

  const dispatch = useDispatch();

  const descriptionRef = useRef(null);

  const [description, setDescription] = useState('');
  const [weekStart, setWeekStart] = useState('');

  const [updatingGroupDetails, setUpdatingGroupDetails] = useState(false);
  const [errorUpdatingGroupDetails, setErrorUpdatingGroupDetails] = useState(false);

  const resetDetails = (groupDetails: Group): void => {
    setDescription(groupDetails.description);
    setWeekStart(groupDetails.week_start);
  };

  useEffect(() => {
    if (group) {
      resetDetails(group);
    }
  }, [group]);

  const onWeekStartChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    if (e.target.checked) {
      setWeekStart(e.target.value);
    }
  };

  const onSubmitDetails = async (): Promise<void> => {
    setUpdatingGroupDetails(true);

    const newGroup: Group = {
      id: group.id,
      group_name: group.group_name,
      group_title: group.group_title,
      grouppic_name: group.grouppic_name,
      description,
      week_start: weekStart
    };

    const updatedGroup = (await dispatch(putGroup(newGroup))) as Group;

    if (updatedGroup) {
      resetDetails(updatedGroup);
    } else {
      setErrorUpdatingGroupDetails(true);
    }

    setUpdatingGroupDetails(false);
  };

  const onCancelDetails = (): void => {
    resetDetails(group);
  };

  return (
    <div className={classes.editGroup}>
      <h3 className={classes.title}>Group Details</h3>
      <div className={classes.form}>
        <div>
          <p className={classes.inputTitle}>Description</p>
          <AutoResizeTextArea
            maxLength={1000}
            placeholder="..."
            onChange={(e: React.ChangeEvent<HTMLTextAreaElement>): void => setDescription(e.target.value)}
            useCustomValue={true}
            value={description}
            disabled={false}
            className={classes.textArea}
            ref={descriptionRef}
          />
        </div>
        <div>
          <p className={classes.inputTitle}>Week Start</p>
          <div className={classes.radioGroup}>
            <RadioButton
              id="sunday"
              name="weekStart"
              value="sunday"
              label="Sunday"
              onChange={onWeekStartChange}
              defaultChecked={group.week_start === 'sunday'}
              className={classes.radio}
            />
            <RadioButton
              id="monday"
              name="weekStart"
              value="monday"
              label="Monday"
              onChange={onWeekStartChange}
              defaultChecked={group.week_start === 'monday'}
              className={classes.radio}
            />
          </div>
        </div>
        <div className={classes.actions}>
          <AJButton
            type="contained"
            disabled={false}
            onClick={onSubmitDetails}
            className={classNames(classes.submitButton, updatingGroupDetails && classes.disabledSubmitButton)}
          >
            <p>{updatingGroupDetails ? 'Saving Details...' : 'Save Details'}</p>
            {updatingGroupDetails && <LoadingSpinner className={classes.buttonSpinner} />}
          </AJButton>
          <AJButton type="text" disabled={false} onClick={onCancelDetails} className={classes.cancelButton}>
            Cancel
          </AJButton>
        </div>
      </div>
      <h3 className={classes.title}>Profile Picture</h3>
      <div className={classes.form}>
        <UploadGroupPicture
          group={group}
          groupPictureUrl={
            group.grouppic_name ? `/uasset/group/${group.id}/${group.grouppic_name}` : '/asset/saintsxctf.png'
          }
        />
      </div>
      {errorUpdatingGroupDetails && (
        <DefaultErrorPopup
          message="Failed to update the group details"
          onClose={(): void => setErrorUpdatingGroupDetails(false)}
          retryable={true}
          onRetry={onSubmitDetails}
        />
      )}
    </div>
  );
};

export default EditGroup;
