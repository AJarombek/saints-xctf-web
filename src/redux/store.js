/**
 * Build and configure the redux store, which holds the application state and handles
 * all state updates.
 * @author Andrew Jarombek
 * @since 5/1/2020
 */

import { createBrowserHistory } from 'history';
import { createStore, applyMiddleware } from 'redux';
import { routerMiddleware } from 'connected-react-router';
import createLogger from 'redux-logger';
import reducer from './modules/reducers';

const history = createBrowserHistory();

const configureStore = (data) => {
  const loggerMiddleware = createLogger();
  const middleware = [loggerMiddleware, routerMiddleware(history)];

  const createStoreWithMiddleware = applyMiddleware(...middleware)(createStore);
  return createStoreWithMiddleware(reducer, data);
};

export { history };
export default configureStore;
