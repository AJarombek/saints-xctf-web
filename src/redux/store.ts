/**
 * Build and configure the redux store, which holds the application state and handles
 * all state updates.
 * @author Andrew Jarombek
 * @since 5/1/2020
 */

import { createBrowserHistory } from 'history';
import { createStore, applyMiddleware, Store, Action } from 'redux';
import { routerMiddleware } from 'connected-react-router';
import { createLogger } from 'redux-logger';
import thunkMiddleware, { ThunkAction } from 'redux-thunk';
import reducer from './modules/reducers';
import { RootState } from './types';

export const history = createBrowserHistory();

export default function configureStore(): Store {
  const loggerMiddleware = createLogger();
  const middleware = [routerMiddleware(history), loggerMiddleware, thunkMiddleware];

  const createStoreWithMiddleware = applyMiddleware(...middleware)(createStore);
  const store = createStoreWithMiddleware(reducer(history));

  if (process.env.NODE_ENV !== 'production' && module.hot) {
    module.hot.accept('./modules/reducers', () => store.replaceReducer(reducer));
  }

  return store;
}

export type AppThunk<ReturnType = void, State = RootState> = ThunkAction<ReturnType, State, unknown, Action<string>>;
