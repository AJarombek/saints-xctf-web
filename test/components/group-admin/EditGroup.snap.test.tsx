/**
 * Snapshot test for the EditGroup component.
 * @author Andrew Jarombek
 * @since 3/5/2021
 */

import React from 'react';
import renderer from 'react-test-renderer';
import { Store } from 'redux';
import { emptyStore } from '../../test-utils/storeMocks';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import EditGroup from '../../../src/components/group-admin/EditGroup';
import { alumni } from '../../test-utils/groupMocks';

const mockStore = configureStore([thunk]);

describe('EditGroup Snapshot Tests', () => {
  let store: Store;

  beforeEach(() => {
    store = mockStore(emptyStore);
  });

  it('renders correctly', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <EditGroup group={alumni} />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
