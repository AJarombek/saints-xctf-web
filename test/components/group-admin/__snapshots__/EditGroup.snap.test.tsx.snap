// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditGroup Snapshot Tests renders correctly 1`] = `
<div
  className="editGroup-0-1-1"
>
  <h3
    className="title-0-1-2"
  >
    Group Details
  </h3>
  <div
    className="form-0-1-3"
  >
    <div>
      <p
        className="inputTitle-0-1-4"
      >
        Description
      </p>
      <textarea
        className="textArea-0-1-13 textArea-0-1-7"
        disabled={false}
        maxLength={1000}
        onChange={[Function]}
        onKeyUp={[Function]}
        placeholder="..."
        value=""
      />
    </div>
    <div>
      <p
        className="inputTitle-0-1-4"
      >
        Week Start
      </p>
      <div
        className="radioGroup-0-1-5"
      >
        <div
          className="radio-0-1-14 radio-0-1-6"
        >
          <label>
            <div
              className="inputWrapper-0-1-15"
            >
              <input
                className="input-0-1-16"
                defaultChecked={true}
                id="sunday"
                name="weekStart"
                onChange={[Function]}
                type="radio"
                value="sunday"
              />
              <div
                className="customRadio-0-1-17"
              />
            </div>
            <div
              className="customLabel-0-1-18"
            >
              Sunday
            </div>
          </label>
        </div>
        <div
          className="radio-0-1-14 radio-0-1-6"
        >
          <label>
            <div
              className="inputWrapper-0-1-15"
            >
              <input
                className="input-0-1-16"
                defaultChecked={false}
                id="monday"
                name="weekStart"
                onChange={[Function]}
                type="radio"
                value="monday"
              />
              <div
                className="customRadio-0-1-17"
              />
            </div>
            <div
              className="customLabel-0-1-18"
            >
              Monday
            </div>
          </label>
        </div>
      </div>
    </div>
    <div
      className="actions-0-1-8"
    >
      <div
        className="ajBaseButton-2-2-4 ajBaseButtonEnabled-2-2-5 aj-contained-button ajContainedButton-2-2-1 ajContainedButtonEnabled-2-2-2 submitButton-0-1-9"
        onClick={[Function]}
      >
        <button
          disabled={false}
          type="button"
        >
          <p>
            Save Details
          </p>
        </button>
      </div>
      <div
        className="ajBaseButton-2-2-4 ajBaseButtonEnabled-2-2-5 aj-text-button ajTextButton-2-2-7 ajTextButtonEnabled-2-2-8 cancelButton-0-1-10"
        onClick={[Function]}
      >
        <button
          disabled={false}
          type="button"
        >
          Cancel
        </button>
      </div>
    </div>
  </div>
  <h3
    className="title-0-1-2"
  >
    Profile Picture
  </h3>
  <div
    className="form-0-1-3"
  >
    <div
      className="pictureContainer-0-1-19"
    >
      <figure
        className="picture-0-1-20"
      >
        <img
          alt=""
          src="/uasset/group/1/picture.png"
        />
      </figure>
      <div
        className="uploadFile-0-1-27 uploadFileBoxEmpty-0-1-29"
        onClick={[Function]}
        onDragEnter={[Function]}
        onDragLeave={[Function]}
        onDragOver={[Function]}
        onDrop={[Function]}
      >
        <input
          accept="image/*"
          className="hiddenInput-0-1-30"
          onChange={[Function]}
          type="file"
        />
        <p
          className="uploadText-0-1-31 uploadTextEmpty-0-1-33"
        >
          Click here or drag and drop a file.
        </p>
      </div>
    </div>
    <div
      className="note-0-1-26"
    >
      <p>
        NOTE: Crop your image into a square for the best results.
      </p>
    </div>
    <div
      className="actions-0-1-21"
    >
      <div
        className="ajBaseButton-2-2-4 ajBaseButtonDisabled-2-2-6 aj-contained-button ajContainedButton-2-2-1 ajContainedButtonDisabled-2-2-3 submitButton-0-1-22 disabledSubmitButton-0-1-23"
        onClick={null}
      >
        <button
          disabled={true}
          type="button"
        >
          <p>
            Save Picture
          </p>
        </button>
      </div>
      <div
        className="ajBaseButton-2-2-4 ajBaseButtonEnabled-2-2-5 aj-text-button ajTextButton-2-2-7 ajTextButtonEnabled-2-2-8 cancelButton-0-1-25"
        onClick={[Function]}
      >
        <button
          disabled={false}
          type="button"
        >
          Cancel
        </button>
      </div>
    </div>
  </div>
</div>
`;
